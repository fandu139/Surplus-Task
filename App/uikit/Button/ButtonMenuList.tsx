import { ICON_ARROW_RIGHT } from '../../assets/icon';
import Colors from '../../theme/colors';
import Icon from '../Icon';
import Text from '../Text';
import React from 'react';
import { View, StyleSheet, TouchableOpacity, StyleProp, ViewStyle, TextStyle } from 'react-native';
import FastImage from 'react-native-fast-image';
import AppStyles from '../../theme/appStyles';
import Divider from '../Divider';

/**
 *
 * @param testID
 * @param accessibilityLabel
 * @param iconLeft
 * @param imageLeft
 * @param iconLeftColor
 * @param iconRight
 * @param iconRightColor
 * @param rightTitle
 * @param title
 * @param onPress
 * @param titleStyle
 * @param style
 * @param useDivider
 * @param useBold
 * @param rightTitleStyle
 * @param headerText
 * @param headerRightText
 * @param description
 * @param disabled
 * @param subtitle
 * @param rightContent
 * @returns {*}
 * @constructor
 */

interface Props {
  testID: string;
  testIDTitle?: string;
  testIDRightTitle?: string;
  accessibilityLabel?: string;
  iconLeft?: string;
  imageLeft?: number;
  iconLeftColor?: string;
  iconRight?: string | null;
  iconRightColor?: string;
  iconRightSize?: number;
  onPressIconRight?: () => void;
  testIDIconRight?: string;
  rightTitle?: string;
  title?: string | React.ReactNode;
  onPress?: () => void;
  titleStyle?: StyleProp<TextStyle>;
  style?: StyleProp<ViewStyle>;
  useDivider?: boolean;
  useBold?: boolean;
  rightTitleStyle?: StyleProp<TextStyle>;
  headerText?: string;
  headerRightText?: string;
  description?: string;
  disabled?: boolean;
  subtitle?: React.ReactNode;
  rightContent?: React.ReactNode;
  rightTitleSelectable?: boolean;
}

const ButtonMenuList: React.FC<Props> = ({
  testID,
  testIDTitle,
  testIDRightTitle,
  accessibilityLabel,
  iconLeft,
  imageLeft,
  iconLeftColor,
  iconRight,
  iconRightColor,
  iconRightSize,
  onPressIconRight,
  testIDIconRight,
  rightTitle,
  title,
  onPress,
  titleStyle,
  style,
  useDivider,
  useBold,
  rightTitleStyle,
  headerText,
  headerRightText,
  description,
  disabled,
  subtitle,
  rightContent,
  rightTitleSelectable,
}: Props) => {
  const isVerticalStyle = !!(headerText && headerRightText && description);
  const Title = useBold ? Text.Bold : Text;

  return (
    <TouchableOpacity
      disabled={disabled}
      activeOpacity={0.7}
      onPress={onPress}
      testID={testID}
      style={[styles.container, style, isVerticalStyle && styles.containerRow]}
      accessibilityLabel={testID || accessibilityLabel}
    >
      {isVerticalStyle &&
        (imageLeft ? (
          <View style={styles.contentImage}>
            <FastImage
              resizeMode={FastImage.resizeMode.contain}
              style={styles.imageItem}
              source={imageLeft as number}
            />
          </View>
        ) : (
          <View style={styles.contentImage}>
            <Icon
              name={iconLeft}
              style={styles.iconLeft}
              size={22}
              color={iconLeftColor as string}
            />
          </View>
        ))}
      <View style={[styles.content, isVerticalStyle && styles.containerColumn]}>
        {isVerticalStyle && (
          <View style={styles.rightTextContainer}>
            <Text style={titleStyle} testID={`title-${testID}`}>
              {headerText}
            </Text>
            <Text style={styles.headerRightText} testID={`subtitle-${testID}`}>
              {headerRightText}
            </Text>
          </View>
        )}
        {!isVerticalStyle && (
          <Icon
            size={18}
            style={styles.iconContainer}
            color={iconLeftColor as string}
            name={iconLeft}
          />
        )}
        {!isVerticalStyle && imageLeft !== 1 && (
          <View style={styles.iconContainer}>
            <FastImage
              resizeMode={FastImage.resizeMode.contain}
              style={styles.imageItem}
              source={imageLeft as number}
            />
          </View>
        )}
        <View style={styles.listContent}>
          <View style={rightContent ? styles.innerListContentFreeStyle : styles.innerListContent}>
            <View style={styles.leftTextContainer}>
              <View>
                {typeof title === 'string' ? (
                  <Title testID={testIDTitle || `title-${testID}`} style={titleStyle}>
                    {title}
                  </Title>
                ) : (
                  title
                )}
                {subtitle}
              </View>
            </View>
            {rightContent || (
              <View style={styles.rightTextContainer}>
                {(!!description || !!rightTitle) && (
                  <Text
                    testID={testIDRightTitle}
                    style={rightTitleStyle}
                    selectable={rightTitleSelectable}
                  >
                    {isVerticalStyle ? description : rightTitle}
                  </Text>
                )}
                {iconRight ? (
                  <Icon
                    style={styles.iconRight}
                    size={iconRightSize as number}
                    name={iconRight}
                    color={iconRightColor as string}
                    onPress={onPressIconRight}
                    testID={testIDIconRight}
                  />
                ) : undefined}
              </View>
            )}
          </View>
        </View>
      </View>
      {useDivider && <Divider color={Colors.GRAYE0} />}
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  container: {
    justifyContent: 'center',
    paddingTop: 20,
  },
  containerRow: {
    ...AppStyles.rowItems,
    justifyContent: 'flex-start',
  },
  containerColumn: {
    flex: 1,
    flexDirection: 'column',
  },
  content: {
    flexDirection: 'row',
  },
  contentImage: {
    justifyContent: 'center',
    paddingRight: 15,
  },
  leftTextContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    flex: 1,
  },
  rightTextContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  iconContainer: {
    marginRight: 10,
  },
  imageItem: {
    width: 36,
    height: 36,
  },
  iconLeft: {
    marginRight: 15,
  },
  iconRight: {
    marginLeft: 15,
  },
  headerRightText: {
    color: Colors.GRAY,
    fontSize: 12,
  },
  listContent: {
    flex: 1,
  },
  innerListContent: {
    ...AppStyles.rowItemsCenterSpace,
    marginBottom: 15,
  },
  innerListContentFreeStyle: {
    ...AppStyles.rowItemsSpace,
    marginBottom: 15,
  },
});

ButtonMenuList.defaultProps = {
  accessibilityLabel: undefined,
  testIDTitle: undefined,
  testIDRightTitle: undefined,
  iconLeftColor: Colors.GREEN_47,
  iconRightColor: Colors.GREEN_47,
  iconRightSize: 12,
  iconLeft: undefined,
  imageLeft: 1,
  iconRight: ICON_ARROW_RIGHT,
  onPressIconRight: undefined,
  testIDIconRight: undefined,
  useDivider: true,
  useBold: true,
  headerText: undefined,
  headerRightText: undefined,
  description: undefined,
  rightContent: undefined,
  disabled: false,
  subtitle: undefined,
  rightTitleStyle: {},
  style: {},
  titleStyle: {},
  rightTitle: undefined,
  title: undefined,
  onPress: () => null,
  rightTitleSelectable: false,
};

export default ButtonMenuList;
