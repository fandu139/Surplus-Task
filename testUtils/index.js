import AuthContext from '../App/context/Auth';
import React from 'react';
import { act, fireEvent, render } from '@testing-library/react-native';

const customRender = (ui, options) => {
  const isAuthenticated = options?.isAuthenticated || false;
  const renderOptions = options?.rendererOptions;
  const authContext = options?.authContext;
  return render(
    <AuthContext.Provider value={{ isAuthenticated, ...authContext }}>{ui}</AuthContext.Provider>,
    renderOptions,
  );
};

// override render method
export { customRender as render };

export function wait(ms) {
  return act(() => new Promise((r) => setTimeout(r, ms)));
}

// Due to debounce on Button UI Kit, we need to add milisecond waiting timer
export function fireAsyncPress(element, ms = 100) {
  fireEvent.press(element);
  return wait(ms);
}
