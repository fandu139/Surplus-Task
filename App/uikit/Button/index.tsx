import React from 'react';
import {
  Pressable,
  StyleSheet,
  TouchableOpacity,
  View,
  StyleProp,
  ViewStyle,
  TextStyle,
} from 'react-native';
import FastImage from 'react-native-fast-image';
import ButtonMenuList from './ButtonMenuList';
import AppStyles from '../../theme/appStyles';
import Fonts from '../../theme/fonts';
import Icon from '../../uikit/Icon';
import Spinner from '../../uikit/Spinner';
import Text from '../../uikit/Text';
import Colors from '../../theme/colors';

/**
 * Documentation: https://reactnative.dev/docs/pressable
 * Based on
 * @param disabled
 * @param loading
 * @param icon
 * @param iconFarRight
 * @param mode
 * @param onPress
 * @param testID
 * @param accessibilityLabel
 * @param style
 * @param textStyle
 * @param title
 * @param useFallback
 * @returns {*}
 * @constructor
 */

type Mode = 'contained' | 'outlined' | 'custom' | 'contained-inverted' | 'outlined-bw';
type StyleConfig =
  | {
      button: StyleProp<ViewStyle>;
      text: StyleProp<TextStyle>;
      buttonPressed: StyleProp<TextStyle>;
      spinner: string;
      icon: string;
    }
  | Record<string, never>;

interface Props {
  title: string | undefined;
  onPress: () => void;
  disabled?: boolean;
  loading?: boolean;
  icon?: {
    name: string;
    size?: number;
    color?: string;
  };
  iconFarRight?: {
    name: string;
    size: number;
    color: string;
  };
  image?: number
    | {
      uri?: string;
      headers?: {
        [key: string]: string;
      };
      priority?: 'low' | 'normal' | 'high';
      cache?: 'immutable' | 'web' | 'cacheOnly';
    };
  mode?: Mode;
  testID: string;
  testIDTitle?: string;
  accessibilityLabel?: string;
  accessibilityLabelTitle?: string;
  style?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
  useFallback?: boolean;
}

interface ButtonProps {
  MenuList: typeof ButtonMenuList;
  Mode: {
    Contained: 'contained';
    Outlined: 'outlined';
    Custom: 'custom';
    ContainedInverted: 'contained-inverted';
  };
}

const Button: React.FC<Props> & ButtonProps = ({
  disabled,
  loading,
  icon,
  iconFarRight,
  mode,
  onPress,
  testID,
  testIDTitle,
  accessibilityLabel,
  accessibilityLabelTitle,
  style,
  textStyle,
  title,
  useFallback,
  image,
}: Props) => {
  let styleConfig: StyleConfig = {};

  if (mode === 'contained') {
    styleConfig = {
      button: [styles.buttonContained, disabled && styles.buttonContainedDisabled],
      text: [styles.textContained, disabled && styles.textContainedDisabled],
      buttonPressed: styles.buttonContainedPressed,
      spinner: Colors.WHITE,
      icon: Colors.WHITE,
    };
  }

  if (mode === 'outlined') {
    styleConfig = {
      button: [styles.buttonOutlined, disabled && styles.buttonOutlinedDisabled],
      text: [styles.textOutlined, disabled && styles.textOutlinedDisabled],
      buttonPressed: styles.buttonOutlinedPressed,
      spinner: Colors.RED,
      icon: Colors.RED,
    };
  }

  if (mode === 'outlined-bw') {
    styleConfig = {
      button: [styles.buttonOutlinedBW, disabled && styles.buttonOutlinedBWDisabled],
      text: [styles.textOutlinedBW, disabled && styles.textOutlinedBWDisabled],
      buttonPressed: styles.buttonOutlinedBWPressed,
      spinner: Colors.GRAY,
      icon: Colors.GRAY,
    };
  }

  if (mode === 'contained-inverted') {
    styleConfig = {
      button: [styles.buttonContainedInverted, disabled && styles.buttonContainedDisabled],
      text: [styles.textContainedInverted, disabled && styles.textContainedInvertedDisabled],
      buttonPressed: styles.buttonContainedInvertedPressed,
      spinner: Colors.RED,
      icon: Colors.RED,
    };
  }

  if (useFallback) {
    return (
      <TouchableOpacity
        activeOpacity={0.4}
        style={[styleConfig.button, style]}
        disabled={disabled}
        testID={testID}
        accessibilityLabel={testID || accessibilityLabel}
        onPress={() => {
          if (!loading) {
            return onPress();
          }
          return null;
        }}
        hitSlop={AppStyles.HitSlopValue}
      >
        {loading ? (
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          <Spinner size={style?.height / 2 || 24} color={styleConfig.spinner} />
        ) : (
          <View style={styles.buttonTextContainer}>
            {icon && (
              <Icon
                style={styles.iconMargin}
                name={icon.name}
                size={icon.size}
                color={icon.color || styleConfig.icon}
              />
            )}
            {image && <FastImage style={styles.image} source={image} />}
            <Text.Black style={[styleConfig.text, textStyle]}>{title}</Text.Black>
          </View>
        )}
        {iconFarRight && (
          <View style={styles.iconFarRight}>
            {/* Ignore TS Ignore on Icon
              // @ts-ignore */}
            <Icon
              name={iconFarRight.name}
              size={iconFarRight.size}
              color={iconFarRight.color || styleConfig.icon}
            />
          </View>
        )}
      </TouchableOpacity>
    );
  }

  return (
    <Pressable
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      delayPressIn={0}
      disabled={disabled}
      testID={testID}
      accessibilityLabel={testID || accessibilityLabel}
      onPress={() => {
        if (!loading) {
          return onPress();
        }
        return null;
      }}
      style={({ pressed }) => [styleConfig.button, pressed && styleConfig.buttonPressed, style]}
      hitSlop={AppStyles.HitSlopValue}
    >
      {loading ? (
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        <Spinner size={style?.height / 2 || 24} color={styleConfig.spinner} />
      ) : (
        <View style={styles.buttonTextContainer}>
          {icon && (
            <Icon
              style={styles.iconMargin}
              name={icon.name}
              size={icon.size}
              color={icon.color || styleConfig.icon}
            />
          )}
          {image && <FastImage style={styles.image} source={image} />}
          <Text.Black
            style={[styleConfig.text, textStyle]}
            testID={testIDTitle}
            accessibilityLabel={accessibilityLabelTitle || testIDTitle}
          >
            {title}
          </Text.Black>
        </View>
      )}
      {iconFarRight && (
        <View style={styles.iconFarRight}>
          <Icon
            name={iconFarRight.name}
            size={iconFarRight.size}
            color={iconFarRight.color || styleConfig.icon}
          />
        </View>
      )}
    </Pressable>
  );
};

const styles = StyleSheet.create({
  buttonContained: {
    backgroundColor: Colors.GREEN_47,
    color: Colors.WHITE,
    borderRadius: 5,
    alignItems: 'center',
    justifyContent: 'center',
    height: 48,
  },
  buttonContainedDisabled: {
    backgroundColor: Colors.GRAYF3,
  },
  buttonContainedPressed: {
    backgroundColor: Colors.RED_PRESSED,
  },
  buttonContainedInverted: {
    backgroundColor: Colors.GRAYF3,
    color: Colors.WHITE,
    borderRadius: 5,
    alignItems: 'center',
    justifyContent: 'center',
    height: 48,
  },
  buttonContainedInvertedDisabled: {
    backgroundColor: Colors.GRAYF3,
  },
  buttonContainedInvertedPressed: {
    backgroundColor: Colors.GRAYF3,
  },
  textContainedInverted: {
    fontSize: Fonts.size.medium,
    color: Colors.GREEN_47,
  },
  textContainedInvertedDisabled: {
    color: Colors.GRAY,
  },
  buttonTextContainer: {
    ...AppStyles.rowItemsCenter,
  },
  iconMargin: {
    marginRight: 5,
  },
  iconFarRight: {
    width: 30,
    height: 30,
    position: 'absolute',
    right: 10,
    alignItems: 'center',
    justifyContent: 'center',
  },
  textContained: {
    fontSize: Fonts.size.medium,
    color: Colors.WHITE,
  },
  textContainedDisabled: {
    color: Colors.GRAY,
  },
  buttonOutlined: {
    borderColor: Colors.RED,
    borderWidth: 1,
    borderRadius: 5,
    alignItems: 'center',
    justifyContent: 'center',
    height: 48,
  },
  buttonOutlinedDisabled: {
    borderColor: Colors.GRAYF3,
  },
  buttonOutlinedPressed: {
    opacity: 0.8,
    borderColor: Colors.RED_PRESSED,
  },
  textOutlined: {
    fontSize: Fonts.size.medium,
    color: Colors.RED,
  },
  textOutlinedDisabled: {
    color: Colors.GRAY,
  },
  buttonOutlinedBW: {
    borderColor: Colors.WHITE,
    borderWidth: 1,
    borderRadius: 5,
    alignItems: 'center',
    justifyContent: 'center',
    height: 48,
  },
  buttonOutlinedBWDisabled: {
    borderColor: Colors.GRAYF3,
  },
  buttonOutlinedBWPressed: {
    opacity: 0.8,
    borderColor: Colors.BLACK_50,
  },
  textOutlinedBW: {
    fontSize: Fonts.size.medium,
    color: Colors.WHITE,
  },
  textOutlinedBWDisabled: {
    color: Colors.GRAY,
  },
  image: {
    width: 20,
    height: 20,
  },
});

Button.defaultProps = {
  disabled: false,
  loading: false,
  icon: undefined,
  iconFarRight: undefined,
  style: {},
  textStyle: {},
  mode: 'contained',
  accessibilityLabel: undefined,
  testIDTitle: undefined,
  accessibilityLabelTitle: undefined,
  useFallback: true,
  image: undefined,
};

Button.Mode = {
  Contained: 'contained',
  Outlined: 'outlined',
  Custom: 'custom',
  ContainedInverted: 'contained-inverted',
};

Button.MenuList = ButtonMenuList;

export default Button;
