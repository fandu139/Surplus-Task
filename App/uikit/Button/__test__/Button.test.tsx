import Button from '../';
import React from 'react';
import { render } from '../../../../testUtils';

describe('Button component test', () => {
  describe('snapshot test', () => {
    it('should render correctly', () => {
      const tree = render(
        <Button title="Test Button Standard" onPress={() => null} testID="button-default" />,
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
