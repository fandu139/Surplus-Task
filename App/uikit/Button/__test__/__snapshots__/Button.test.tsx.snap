// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component test snapshot test should render correctly 1`] = `
<View
  accessibilityLabel="button-default"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  hitSlop={
    {
      "bottom": 15,
      "left": 15,
      "right": 15,
      "top": 15,
    }
  }
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#47bba5",
      "borderRadius": 5,
      "color": "#FFFFFF",
      "height": 48,
      "justifyContent": "center",
      "opacity": 1,
    }
  }
  testID="button-default"
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        [
          {
            "color": "#202020",
            "fontSize": 14,
            "fontWeight": "normal",
          },
          [
            [
              {
                "color": "#FFFFFF",
                "fontSize": 16,
              },
              false,
            ],
            {},
          ],
          undefined,
          undefined,
        ]
      }
    >
      Test Button Standard
    </Text>
  </View>
</View>
`;
